### ISSUE TITLE
Should we include an AddOrUpdate() method?

### ISSUE THREAD

**Issue by jtreuting** [SharpRepository/SharpRepository#120]
I've been seeing these more and more recently it seems, many times they are called InsertOrUpdate.  Might be a nice feature to add.

**Comment by mgmccarthy** [comment]
@jtreuting, are you still accepting PR's for this repository (I see 4 outstanding PR's, the oldest one being from 2013)? 

If so, adding AddOrUpdate for EF repository should be straight-forward, and I think for RavenDb as well. Unfortunately, I'm not as familiar with MongoDb, CouchDb or Azure. Also, EF's implementation of AddOrUpdate will lose data if you don't have the value for all properties on the entity on which your calling AddOrUpdate (aka, a true a "upsert"). 

Just wondering if your definition of AddOrUpdate or InsertOrUpdate follows the definition of how EF does it, or if you're proposing a check first for existence, and then an update to the supplied properties.

**Comment by fiorebat** [comment]
just for precision 4 PR are opened, others are accepted, another my PR was accepted on September '15.
As you can see he is not replying in last 2 weeks

**Comment by jtreuting** [comment]
Yes I am, but unfortunately have been overly busy at work and haven't had much time to actually do it.  If I don't find in the next few weeks I'll need to seriously consider enlisting help and giving others access to handle pull requests as well.

**Comment by jtreuting** [comment]
@mgmccarthy  I was originally thinking like EF does it but would be open to alternatives if they make sense.  Thoughts?

**Comment by mgmccarthy** [comment]
@jtreuting, Entity Framework does do this via System.Data.Entity.Migrations.DbSetMigrationsExtensions.AddOrUpdate. Again, I'd have to look a bit more into if RavenDb support something like this. I haven't spent enough time looking around SharpRepository code to know it that well, but would adding AddOrUpdate to only EfRepositoryBase be useful at all (or even possible?). It looks like I should not add AddOrUpdate  to LinqRepositoryBase b/c we would need implementation for all repo types.

As far as other alternatives for the other provider-specific repos, we could go first with:
- check to see if record exists
- if the record does exist, update only the properties provided via the AddOrUpdate to the data store. In other words, don't make it a update that will overwrite all properties on the record regardless of whether or  not all properties are set to values we'd want to update to (this is default entity framework .AddOrupdate behavior).
- if the record does not exist, insert it.

I don't like this approach, as it's chatty, and there would most likely be performance implications, but if you want to implement AddOrUpdate at the LinqRepositoryBase level, it might be a temporary evil until I can figure out the RavenDb repo, and maybe other developers can chip in on the other repos that represent database providers which I no experience with.
